tests/infra/proposal_generator.py

```diff
@@ -0,0 +1,334 @@
+TRIVIAL_NO_BALLOT = {"text": "return false"}
+
+LUA_FUNCTION_EQUAL_ARRAYS = """
+function equal_arrays(a, b)
+  if #a ~= #b then
+    return false
+  else
+    for k, v in ipairs(a) do
+      if b[k] ~= v then
+        return false
+      end
+    end
+    return true
+  end
+end
+"""
+
+
+PROPOSAL_ID_PLACEHOLDER = "<replace with desired proposal_id>"
+
+
+def add_arg_construction(lines, arg, arg_name="args"):
+    if isinstance(arg, list):
+        lines.append(f"{arg_name} = {list_as_lua_literal(arg)}")
+    elif isinstance(arg, dict):
+        lines.append(f"{arg_name} = {{}}")
+        for k, v in args.items():
+            add_arg_construction(lines, v, arg_name=f"{arg_name}.{k}")
+    elif isinstance(arg, str):
+        lines.append(f'{arg_name} = "{arg}"')
+    else:
+        lines.append(f"{arg_name} = {arg}")
+
+
+def add_arg_checks(lines, arg, arg_name="args"):
+    lines.append(f"if {arg_name} == nil then return false")
+    if isinstance(arg, list):
+        expected_name = arg_name.replace(".", "_")
+        lines.append(f"{expected_name} = {list_as_lua_literal(arg)}")
+        lines.append(
+            f"if not equal_arrays({arg_name}, {expected_name}) then return false end"
+        )
+    elif isinstance(arg, dict):
+        for k, v in arg.items():
+            add_arg_checks(lines, v, arg_name=f"{arg_name}.{k}")
+    elif isinstance(arg, str):
+        lines.append(f'if not {arg_name} == "{arg}" then return false end')
+    else:
+        lines.append(f"if not {arg_name} == {arg} then return false end")
+
+
+def build_proposal(proposed_call, args=None, inline_args=False):
+    LOG.debug(f"Generating {proposed_call} proposal")
+
+    proposal_script_lines = []
+    if args is None:
+        proposal_script_lines.append(f'return Calls:call("{proposed_call}")')
+    else:
+        if inline_args:
+            add_arg_construction(proposal_script_lines, args)
+        else:
+            proposal_script_lines.append("tables, args = ...")
+        proposal_script_lines.append(f'return Calls:call("{proposed_call}", args)')
+
+    proposal_script_text = "; ".join(proposal_script_lines)
+    proposal = {
+        "script": {"text": proposal_script_text},
+    }
+    if args is not None and not inline_args:
+        proposal["parameter"] = args
+
+    vote_lines = [
+        "tables, calls = ...",
+        "if not #calls == 1 then return false end",
+        "call = calls[1]",
+        f'if not call.func == "{proposed_call}" then return false end',
+        LUA_FUNCTION_EQUAL_ARRAYS,
+    ]
+    if args is not None:
+        add_arg_checks(vote_lines, args)
+    vote_lines.append("return true")
+    vote_text = "; ".join(vote_lines)
+    vote = {
+        "ballot": {"text": vote_text},
+        "id": PROPOSAL_ID_PLACEHOLDER,
+    }
+
+    LOG.trace(f"Made {proposed_call} proposal:\n{json.dumps(proposal, indent=2)}")
+    LOG.trace(f"Accompanying vote:\n{json.dumps(vote, indent=2)}")
+
+    return proposal, vote
+
+
+class Proposals:
```

Comment:
Why a class with static methods? infra.proposals with functions seems clearer?

Reply:
We generate the CLI commands/arguments using `inspect`, so I thought it was cleaner to have a single class holding the "functions to expose" to introspect rather than the entire module? They could instead get a decorator, or maybe there's some other Pythonic way of doing this? Or maybe I shouldn't be so lazy and we should populate the CLI explicitly...

Reply:
A registration decorator would be nice, works irrespective of the scope. A module is more accurate than a class, because there's no state, and it's just as easy to use inspect.getmembers() on (or .__dict__.items() really...).

```
def proposal(fun):
  cli.register(fun, description=fun.__doc__)

@proposal
def thingie(...):
  """
  Some doc about the proposal
  """
```

Reply:
Replaced this class with a decorator, works great